Source/control.cpp

```diff
@@ -267,11 +267,33 @@ void PrintInfo(const Surface &out)
 
 	auto infoStringLines = (int)std::count(InfoString.str().begin(), InfoString.str().end(), '\n');
 
-	Rectangle line { GetMainPanel().position + Displacement { 177, LineStart[infoStringLines + pnumlines] }, { 288, 12 } };
+	int panelLines = pnumlines + infoStringLines;
+
+	std::string infoStringOut = std::string(InfoString.str());
+
+	if (panelLines >= 4) {
+		panelLines = 3;
+		pnumlines = panelLines - infoStringLines;
+		Log("PrintInfo unable to render everything - not enough lines");
+	}
+
+	Rectangle line { GetMainPanel().position + Displacement { 177, LineStart[panelLines] }, { 288, 12 } };
 
 	if (!InfoString.empty()) {
-		int lineHeight = LineHeights[pnumlines + infoStringLines];
-		DrawString(out, InfoString, line, InfoColor | UiFlags::AlignCenter | UiFlags::KerningFitSpacing, 2, lineHeight);
+		const int lineHeight = LineHeights[panelLines];
+		if (infoStringLines >= 4) {
+			infoStringLines = 3;
+			infoStringOut = "";
+			int i = 0;
+			for (string_view s : SplitByChar(InfoString, '\n')) {

```

Comment:
You can avoid having the `InfoStringOut` variable entirely, eliminating all these allocations on every frame, by simply doing this:


```c++
for (size_t end = 0; end < 3; ++end) {
  pos = InfoString.find('\n', end);
}
DrawString(out, InfoString.str().substr(0, pos), ...);
```

It'd be better to do this on assignment to InfoString rather than during rendering, so that the warning is only logged once (also more efficient).

Reply:
I took logging this only when the `InfoString` changes to mean overload the string_or_view = operator to validate the length of the string there, and print the info. I still get tons of logs about it though, it looks like it's still calling it on every frame.

Reply:
I realize we probably don't want to change the = operator, since we probably need string_or_view for things besides the info panel :smile: 
What did you actually have in mind?

Reply:
Ah, just saw this, please see my new comment